### PR TITLE
Create Trip and TripDay classes.

### PR DESCRIPTION
**Pull request by eshika**
In this CL, I have created the basic Trip and TripDay classes, which will be used to store information about a user's trips.

The Trip class has the following fields: tripName, startDate, endDate, numDays, and tripDays (an ArrayList of TripDay objects). I've created two constructors for Trip, one where the trip is limited to one day and the endDate is the same as the startDate (for the MVP), and one where all the fields can be set (for beyond the MVP).

The TripDay class has the following fields: origin, destination, and locations (an ArrayList of place IDs, which are strings). This is the basic information needed for generating a DirectionsRequest with the Google Maps API. 

Some notes on how the structure of these will change in the future:

1. The TripDay class will also have a field that contains an ArrayList of events (for the calendar), which I can add after Adam's CL #16 is approved and merged.
2. The Trip class will also have a user field that stores the user whose trip it is. Will work with Chris to add this.
3. The TripDay will also have fields for storing the resulting route and information calculated by the Maps API, but I'm still exploring the Maps API's Java client (for calculating routes in the Java servlets instead of the Javascript) to figure out how these can be stored.

### CODE REVIEW

File: src/main/java/com/google/sps/Trip.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    if (tripDays == null) {
+      throw new IllegalArgumentException("tripDays cannot be null. Use empty array instead.");
+    }
+
+    this.tripName = tripName;
+    this.startDate = startDate;
+    this.endDate = endDate;
+    this.numDays = numDays;
```
Comment (adamfweidman) [inline src/main/java/com/google/sps/Trip.java:58]: I suggest adding a check that number of days cannot be less than 1 and that there is also a max, say one month (30 days)

File: src/main/java/com/google/sps/Trip.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+   * @param startDate The start date for the trip. Must be non-null.
+   * @param tripDays The list of tripDays. Must be non-null.
+   */
+  public Trip(String tripName, String startDate, ArrayList<TripDay> tripDays) {
+    if (tripName == null) {
+      throw new IllegalArgumentException("tripName cannot be null");
+    }
+
+    if (startDate == null) {
+      throw new IllegalArgumentException("startDate cannot be null");
```
Comment (adamfweidman) [inline src/main/java/com/google/sps/Trip.java:77]: I think is the one who is checking this but is it possible that the date won't be the right format? like a "65938123". That would be something we need to check against if users can input that

File: src/main/java/com/google/sps/Trip.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    if (tripDays == null) {
+      throw new IllegalArgumentException("tripDays cannot be null. Use empty array instead.");
```
Comment (adamfweidman) [inline src/main/java/com/google/sps/Trip.java:81]: Good error message!

File: src/main/java/com/google/sps/Trip.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    this.tripDays = new ArrayList<>();
+    this.tripDays.addAll(tripDays);
+  }
+
+  /**
+   * Returns the human-readable name for this trip.
+   */
+  public String getTripName() {
+    return tripName;
```
Comment (adamfweidman) [inline src/main/java/com/google/sps/Trip.java:99]: Personal preference: I like to use this.tripName which just ensures that its the global variable. Your choice!

File: src/main/java/com/google/sps/Trip.java (hunk 1 above)
Comment (zpChris) [reply src/main/java/com/google/sps/Trip.java:58]: Building on this, this may be excess / unnecessary, but I'm wondering if you could add a check to determine whether the `numDays` value is consistent with the `startDate` and `endDate` variables. This may be difficult given that those are Strings though, so I'll leave this up to whatever you decide.

File: src/main/java/com/google/sps/Trip.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+/**
+ * Trip is the class for storing a single trip (could be multiple days).
+ */
+public class Trip {
+  private String tripName;
+  private String startDate;
+  private String endDate;
+  private int numDays;
+  private ArrayList<TripDay> tripDays;
```
Comment (zpChris) [inline src/main/java/com/google/sps/Trip.java:27]: nit: Personal preference from my CIS classes, I prefer a broader interface like `List<TripDay>` to be used here. However, I couldn't find anything in the style guide supporting this, so any action is up to you.

Comment (zpChris) [reply src/main/java/com/google/sps/Trip.java:27]: (This same nit is present in different places, namely the parameters of different functions, but I don't want to saturate the review with nits so I'll just leave a single nit here.)

File: src/main/java/com/google/sps/Trip.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    if (tripDays == null) {
+      throw new IllegalArgumentException("tripDays cannot be null. Use empty array instead.");
+    }
+
+    this.tripName = tripName;
+    this.startDate = startDate;
+    this.endDate = endDate;
+    this.numDays = numDays;
+    this.tripDays = new ArrayList<>();
+    this.tripDays.addAll(tripDays);
```
Comment (zpChris) [inline src/main/java/com/google/sps/Trip.java:60]: Is there a reason you don't just set `this.tripDays = tripDays`? Are you worried about accidentally modifying the original parameter, since it doesn't make a copy?

File: src/main/java/com/google/sps/Trip.java (hunk 5 above)
Comment (zpChris) [inline src/main/java/com/google/sps/Trip.java:27]: No action required, just a note. Eventually, I think we'll add a `String tripId` variable that holds a 16-digit alphanumeric identifier, so that we can easily store and access different trips in datastore. Again, no action required on this, just a note for later plans.

File: src/main/java/com/google/sps/Trip.java (hunk 2 above)
Comment (zpChris) [reply src/main/java/com/google/sps/Trip.java:77]: That's true, I think there are ways to do this! You could [check if the String is a valid date](https://stackoverflow.com/questions/33968333/how-to-check-if-a-string-is-date), and if it is, [convert it to Date format](https://www.javatpoint.com/java-string-to-date) in order to confirm the `startDate` and `endDate` values match the value of `numDays` and the size of `tripDays`.

File: src/main/java/com/google/sps/TripDay.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+import java.util.ArrayList;
+
+/**
+ * TripDay objects store the trip/route information for a single day.
+ */
+public class TripDay {
+  private String origin;
+  private String destination;
+  private ArrayList<String> locations;
```
Comment (zpChris) [inline src/main/java/com/google/sps/TripDay.java:25]: Two questions:
1. Will these variables always store Strings, or are you going to later use the Google Maps Java Client to store place objects here? (I just saw in the `TripDayTest` class that these Strings are seemingly **place IDs** -- if this is the case, can you add a comment somewhere explaining this?)
2. Say the last location (POI) is also where the user wants to end their day. Is it included in both `locations` and as the `destination`, or is it no longer included at the end of `locations`?

Comment (zpChris) [reply src/main/java/com/google/sps/TripDay.java:25]: Expanding on this; on the frontend, I haven't provided users any options for putting a `destination` different than their `origin`, because at least for the MVP I assumed this would be the same. Was this a fair assumption, or should I include a spot for where they end their day as well?

My main concern here is that this could become overly complicated once we start to add multiple days -- I'd have to dynamically allow adding different `origin` and `destination` values for every day.

File: src/test/java/com/google/sps/TripTest.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  @Test(expected = IllegalArgumentException.class)
+  public void testTripConstructorMultiDayNull() {
+    Trip trip = new Trip(TRIP_NAME, START_DATE, END_DATE, NUM_DAYS, null);
+  }
+
+  // Test single day constructor with null parameter
+  @Test(expected = IllegalArgumentException.class)
+  public void testTripConstructorSingleDayNull() {
+    Trip trip = new Trip(TRIP_NAME, START_DATE, null);
+  }
```
Comment (zpChris) [inline src/test/java/com/google/sps/TripTest.java:92]: If you end up adding code to verify that the dates are valid and that they match the other inputs, can we also add tests checking this functionality as well? (Only applicable if you end up adding that initial code, of course.)

File: src/main/java/com/google/sps/TripDay.java (hunk 7 above)
Comment (eshika) [reply src/main/java/com/google/sps/TripDay.java]: 1. I have these as Strings (for Place IDs) for now, but this could change depending on how I end up using the Java Client! Still exploring/learning about the Client so will keep you posted on that. Have added a comment explaining that these params are Place ID Strings for now!

2. If the last POI is the last stop on the route, then that will be stored only in destination, not locations. For the MVP, I am just assuming that the origin and destination are the same, so this should not be an issue.

For the MVP, I think the form is fine with just having the user input an origin because we are assuming that all trips will be round trips to and from that location, so I can just set the destination to the origin when calling the DirectionsQuery. Beyond the MVP, we can talk about how we want it to look, but I don't think there is any immediate action required for now.

File: src/main/java/com/google/sps/Trip.java (hunk 6 above)
Comment (eshika) [reply src/main/java/com/google/sps/Trip.java:78]: Yes! That was the reasoning for this, added a comment in the code to clarify :)

File: src/main/java/com/google/sps/Trip.java (hunk 1 above)
Comment (eshika) [reply src/main/java/com/google/sps/Trip.java:78]: Added a check to ensure the number of days is between 1 and 31 (inclusive), and am now computing the number of days using the SimpleDateFormat library so that it is consistent with the startDate and endDate.

File: src/main/java/com/google/sps/Trip.java [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Date;
```
Comment (bheff-work) [inline src/main/java/com/google/sps/Trip.java:19]: Since Java 8 came out, the preference is to use the java.time libraries, not java.util for working with time. Oracle themselves recommend this (https://docs.oracle.com/javase/tutorial/datetime/iso/legacy.html).

Let me know if you have any questions about it!

File: src/main/java/com/google/sps/Trip.java [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+  /**
+   * Creates a new Trip that is only one day by default. 
+   * This constructor will be used for the MVP.
+   * 
+   * @param tripName The human-readable name for the trip. Must be non-null.
+   * @param startDate The start date for the trip. Must be non-null.
+   * @param tripDays The list of tripDays. Must be non-null.
+   */
+  public Trip(String tripName, String startDate, List<TripDay> tripDays) {
```
Comment (bheff-work) [inline src/main/java/com/google/sps/Trip.java:93]: I believe you can call your other constructor from here, which will reduce a lot of the code duplication.

public Trip(String tripName, String startDate, List<TripDay> tripDays) {
  this(tripName, startDate, startDate, tripDays);
}

File: src/main/java/com/google/sps/TripDay.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+   */
+  public String getDestination() {
+    return this.destination;
+  }
+
+  /**
+   * Returns an List<String> of locations for this TripDay.
+   */
+  public List<String> getLocations() {
+    return this.locations;
```
Comment (bheff-work) [inline src/main/java/com/google/sps/TripDay.java:75]: same comment as in the Trip class, you'll want to make a copy of this list if you don't want it to be modified

File: src/test/java/com/google/sps/TripDayTest.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    Assert.assertEquals(tripDay.getOrigin(), HOTEL_ID);
+    Assert.assertEquals(tripDay.getDestination(), HOTEL_ID);
+    Assert.assertEquals(tripDay.getLocations(), locations);
+  }
+
+  // Test TripDay constructor with null parameter
+  @Test(expected = IllegalArgumentException.class)
+  public void testTripDayConstructorNull() {
+    TripDay tripDay = new TripDay(HOTEL_ID, HOTEL_ID, null);
+  }
```
Comment (bheff-work) [inline src/test/java/com/google/sps/TripDayTest.java:55]: It would be a good idea to test all the expected failures

File: src/test/java/com/google/sps/TripTest.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+package com.google.sps;
+
+import com.google.sps.Trip;
+import com.google.sps.TripDay;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import java.util.ArrayList;
+import java.util.Arrays;
```
Comment (bheff-work) [inline src/test/java/com/google/sps/TripTest.java:24]: nit: order your imports alphabetically

File: src/test/java/com/google/sps/TripTest.java [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    
+    tripDays = new ArrayList<>();
+    tripDays.add(tripDay);
+
+    Trip trip = new Trip(TRIP_NAME, "06-28-2020", END_DATE, tripDays);
+  }
+
+  // Test multiple day constructor with over a month
+  @Test(expected = IllegalArgumentException.class)
+  public void testTripConstructorMultiDayLongTrip() {
```
Comment (bheff-work) [inline src/test/java/com/google/sps/TripTest.java:100]: this is great, since it captures the edge case of being at 31 days, but you also need a test for the edge case of it's '0'. That test would have caught your off-by-one bug in Trip. Whenever you have restrictions like that in your code, it's important to test right on the edge of those restrictions

File: src/test/java/com/google/sps/TripTest.java [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+    tripDays.add(tripDay);
+
+    Trip trip = new Trip(TRIP_NAME, START_DATE, tripDays);
+
+    Assert.assertEquals(trip.getTripName(), TRIP_NAME);
+    Assert.assertEquals(trip.getStartDate(), START_DATE);
+    Assert.assertEquals(trip.getEndDate(), START_DATE);
+    Assert.assertEquals(trip.getNumDays(), 1);
+    Assert.assertEquals(trip.getTripDays(), tripDays);
+
```
Comment (bheff-work) [inline src/test/java/com/google/sps/TripTest.java:78]: nit: remove extra line

File: src/test/java/com/google/sps/TripTest.java [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+  // Test multiple day constructor with wrong date format
+  @Test(expected = IllegalArgumentException.class)
+  public void testTripConstructorMultiDayInvalidDate() {
+    TripDay tripDay = new TripDay(HOTEL_ID, HOTEL_ID, locations);
+    
+    tripDays = new ArrayList<>();
+    tripDays.add(tripDay);
+
+    Trip trip = new Trip(TRIP_NAME, "06-28-2020", END_DATE, tripDays);
+  }
```
Comment (bheff-work) [inline src/test/java/com/google/sps/TripTest.java:96]: test invalid end date too

File: src/main/java/com/google/sps/Trip.java [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+import java.text.SimpleDateFormat;
+
+/**
+ * Trip is the class for storing a single trip (could be multiple days).
+ */
+public class Trip {
+  private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd");
+
+  private String tripName;
+  private String startDate;
```
Comment (bheff-work) [inline src/main/java/com/google/sps/Trip.java:30]: Perhaps it would be best to keep the date fields in an actual time-based object? That way people wouldn't have to convert it themselves and it makes it easier to display however they would like?

**Comment by eshika** [comment]
Few notes:
1. Per Adam's comment, I have added a constant 31 day trip limit to multi day trips, but this can be increased/modified later as needed.
2. Per Bootsie's comment, I have switched to using the LocalDate library for storing Trip StartDate and EndDate and calculating the number of days in a trip.

File: src/main/java/com/google/sps/Trip.java [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+
+  /**
+   * Creates a new Trip.
+   *
+   * @param tripName The human-readable name for the trip. Must be non-null.
+   * @param startDate The start date for the trip. Must be non-null. Must be in yyyy-MM-dd date format.
+   * @param endDate The end date for the trip. Must be non-null. Must be in yyyy-MM-dd date format.
+   * @param tripDays The list of tripDays. Must be non-null.
+   */
+  public Trip(String tripName, String startDate, String endDate, List<TripDay> tripDays) {
```
Comment (bheff-work) [inline src/main/java/com/google/sps/Trip.java:45]: Why not take in LocalDateTime objects as startDate and endDate? Or at least have another constructor that does?

File: src/main/java/com/google/sps/Trip.java [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+
+  /**
+   * Returns the number of days in this trip.
+   */
+  public int getNumDays() {
+    return this.numDays;
+  }
+
+  /**
+   * Returns an List<TripDay> copy of TripDays for this trip.
```
Comment (bheff-work) [inline src/main/java/com/google/sps/Trip.java:126]: nit: replace 'an' with 'a'

File: src/main/java/com/google/sps/Trip.java [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+
+    if (tripDays == null) {
+      throw new IllegalArgumentException("tripDays cannot be null. Use empty array instead.");
+    }
+
+    if (!isValidDate(startDate)) {
+      throw new IllegalArgumentException("Invalid startDate format. Must be in yyyy-MM-dd date format.");
+    }
+
+    if (!isValidDate(endDate)) {
```
Comment (bheff-work) [inline src/main/java/com/google/sps/Trip.java:66]: not sure you need these methods anymore, since you're doing the LocalDate.parse() below. Maybe have the method check *and* create the LocalDate? That way you aren't having to parse twice, and you don't have to have two try catches in the constructor

File: src/main/java/com/google/sps/Trip.java [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+    return true;
+  }
+
+  /**
+   * Calculates number of days between two dates.
+   * @param startDate The LocalDate representation of start date.
+   * @param endDate The LocalDate representation of end date.
+   */
+  private static int calcNumDays(LocalDate startDate, LocalDate endDate) {
+	  int numDays = ((int) ChronoUnit.DAYS.between(startDate, endDate)) + 1;
```
Comment (bheff-work) [inline src/main/java/com/google/sps/Trip.java:153]: indentation is off here

File: src/test/java/com/google/sps/TripTest.java (hunk 14 above)
Comment (bheff-work) [reply src/test/java/com/google/sps/TripTest.java:150]: I still don't see a test that checks for trip of length 0? Am I missing it?

File: src/main/java/com/google/sps/Trip.java (hunk 18 above)
Comment (eshika) [reply src/main/java/com/google/sps/Trip.java:45]: The user input form gives us the dates as Strings. Since we'll be using the user input to construct the Trips, I chose to design the constructor this way.

File: src/test/java/com/google/sps/TripTest.java (hunk 14 above)
Comment (eshika) [reply src/test/java/com/google/sps/TripTest.java:150]: Good catch! Because a trip with the same start and end date is considered to have 1 day, a trip of length 0 would occur when the end date is one day before the start date. I added a test for this to ensure that the IllegalArgumentException is thrown.
